### PR TITLE
Clean adb forward resource when kill chromedriver process

### PR DESCRIPTION
**Pull request by babyrjw**
Appium start a webview session by establishing a connection between ADB with chromedriver. It kill all of chromedriver processes existed. After establishing a connection ADB will listen on a random tcp port and chromedriver will connect this port. This port was been listening after chromedriver process has been killed. The quantity of ports listened by adb will increased one by one until exhausted if switch context to webview frequently.

So I add some code into appium-chromedriver/lib/chromedriver.js file

### CODE REVIEW

File: lib/chromedriver.js [hunk 1]
```diff
@@ -233,6 +236,12 @@ class Chromedriver extends events.EventEmitter {
     } catch (err) {
       log.info("No old chromedrivers seemed to exist");
     }
+    try {
+      await (B.promisify(cp.exec))(cmd_adb);
```
Comment (imurchie) [inline lib/chromedriver.js]: Let's move the `B.promisify` step outside, so it only happens once...
```js
const cp_exec = B.promisify(cp.exec);
log.debug(`Killing any old chromedrivers, running: ${cmd}`);
try {
  await cp_exec(cmd);
  log.debug("Successfully cleaned up old chromedrivers");
} catch (err) {
  log.warn("No old chromedrivers seemed to exist");
}

log.debug(`Cleaning any old chromedriver port forwarding resources, running: ${cmd_adb}`);
try {
  await cp_exec(cmd_adb);
  log.debug("Successfully cleaned up adb forward resource");
} catch (err) {
  log.warn("No adb forward resource seemed to exist");
}
```

File: lib/chromedriver.js [hunk 2]
```diff
@@ -216,15 +216,18 @@ class Chromedriver extends events.EventEmitter {
 
   async killAll () {
     let cmd;
+    let cmd_adb;
```
Comment (imurchie) [inline lib/chromedriver.js]: We use camel case for variables. Something like `cmdAdb` or `cmdPortForward`.

File: lib/chromedriver.js (hunk 1 above)
Comment (babyrjw) [inline lib/chromedriver.js]: Thanks for you review. Already changed.

File: lib/chromedriver.js (hunk 2 above)
Comment (babyrjw) [inline lib/chromedriver.js]: Thanks for you suggestion. Already changed.

**Comment by babyrjw** [comment]
Thanks for your review. I have made a change by following your sugguestion

File: lib/chromedriver.js [hunk 3]
```diff
@@ -216,22 +216,33 @@ class Chromedriver extends events.EventEmitter {
...
             "FOR /F \"usebackq\" %b in (`TASKLIST /FI \"PID eq %a\" ^| " +
             "findstr /I chromedriver.exe`) do (IF NOT %b==\"\" TASKKILL " +
             "/F /PID %a))";
+      cmdAdb = "FOR /F \"usebackq tokens=2\" %a in (`adb forward --list ^| findstr /R /C:\"webview\"`) do ( adb forward --remove %a )";
     } else {
       cmd = `pkill -15 -f "${this.chromedriver}.*--port=${this.proxyPort}"`;
+      cmdAdb = "for a in `adb forward --list | grep \"webview\" | awk '{print $2}'` ; do  adb forward --remove $a ; done";
     }
-    log.info(`Killing any old chromedrivers, running: ${cmd}`);
+    const cp_exec = B.promisify(cp.exec);
```
Comment (imurchie) [inline lib/chromedriver.js]: `cp_exec` should also be camel case (`cpExec`).

File: lib/chromedriver.js [hunk 4]
```diff
@@ -216,22 +216,33 @@ class Chromedriver extends events.EventEmitter {
...
-      log.info("Successfully cleaned up old chromedrivers");
+      await cp_exec(cmd);
+      log.debug("Successfully cleaned up old chromedrivers");
     } catch (err) {
-      log.info("No old chromedrivers seemed to exist");
+      log.warn("No old chromedrivers seemed to exist");
+    }
+    log.debug(`Cleaning any old chromedriver port forwarding resources, running: ${cmd_adb}`);
+    try {
+      await cp_exec(cmd_adb);
```
Comment (imurchie) [inline lib/chromedriver.js]: `cmdAdb`

File: lib/chromedriver.js [hunk 5]
```diff
@@ -216,22 +216,33 @@ class Chromedriver extends events.EventEmitter {
...
     try {
-      await (B.promisify(cp.exec))(cmd);
-      log.info("Successfully cleaned up old chromedrivers");
+      await cp_exec(cmd);
+      log.debug("Successfully cleaned up old chromedrivers");
     } catch (err) {
-      log.info("No old chromedrivers seemed to exist");
+      log.warn("No old chromedrivers seemed to exist");
+    }
+    log.debug(`Cleaning any old chromedriver port forwarding resources, running: ${cmd_adb}`);
```
Comment (imurchie) [inline lib/chromedriver.js]: `cmdAdb`

File: lib/chromedriver.js [hunk 6]
```diff
@@ -216,22 +216,33 @@ class Chromedriver extends events.EventEmitter {
...
       cmd = `pkill -15 -f "${this.chromedriver}.*--port=${this.proxyPort}"`;
+      cmdAdb = "for a in `adb forward --list | grep \"webview\" | awk '{print $2}'` ; do  adb forward --remove $a ; done";
     }
-    log.info(`Killing any old chromedrivers, running: ${cmd}`);
+    const cp_exec = B.promisify(cp.exec);
+    log.debug(`Killing any old chromedrivers, running: ${cmd}`);
     try {
-      await (B.promisify(cp.exec))(cmd);
-      log.info("Successfully cleaned up old chromedrivers");
+      await cp_exec(cmd);
```
Comment (jlipps) [inline lib/chromedriver.js]: this will work, but it will not use any of the machinery we have around use of `adb` via the `appium-adb` project. are we sure it will be compatible with all of the assurances we have from using that project? (for example, finding adb at various paths).

i guess one option could be to pass an instance of `ADB` into the constructor, which could then be used in this method.

Comment (imurchie) [inline lib/chromedriver.js]: Hmm. True. We already pass in the adb port.

Comment (babyrjw) [inline lib/chromedriver.js]: Good, I'll change the cmd to use adbPort parameter

Comment (babyrjw) [inline lib/chromedriver.js]: The adbPort past in constructor parameters is not the adb forward port that ADB listened on. It seems that chromedriver will ask ADB to use an other random port. 
I use appium-adb to find ADB binary path.

**Comment by babyrjw** [comment]
Can those changes been approved  ?

File: lib/chromedriver.js [hunk 7]
```diff
@@ -29,6 +30,8 @@ class Chromedriver extends events.EventEmitter {
     this.jwproxy = new JWProxy({server: this.proxyHost, port: this.proxyPort});
     this.verbose = verbose;
     this.logPath = logPath;
+    this.adb = new ADB();
```
Comment (babyrjw) [inline lib/chromedriver.js]: Now I use appium-adb to clean adb forward resources. But appium-adb still has no fowardList method. I think I should pull a request to add that method. :-P

File: lib/chromedriver.js [hunk 8]
```diff
@@ -223,35 +216,35 @@ class Chromedriver extends events.EventEmitter {
 
   async killAll () {
     let cmd;
-    let cmdAdb;
+	let adb = this.adb;
```
Comment (imurchie) [inline lib/chromedriver.js]: No need for this. Just use `this.adb.forwardList` later.

File: lib/chromedriver.js [hunk 9]
```diff
@@ -223,35 +216,35 @@ class Chromedriver extends events.EventEmitter {
...
     }
-    log.debug(`Cleaning any old chromedriver port forwarding resources, running: ${cmdAdb}`);
-    try {
-      await cpExec(cmdAdb);
-      log.debug("Successfully cleaned up adb forward resource");
-    } catch (err) {
-      log.warn("No adb forward resource seemed to exist");
+    log.debug(`Clean any old adb forwarded port socket connections`);
+    let forwardList = await adb.forwardList();
+    if(forwardList){
```
Comment (imurchie) [inline lib/chromedriver.js]: `map` is an odd way to handle this. I would use a `for ... of` loop:
```js
let connections = await this.adb.forwardList();
for (let conn of connections.split('\n')) {
  if (conn.indexOf('webview_devtools') !== -1) {
    let params = conn.split(/\s+/);
    if (params.length > 1) {
      await this.adb.removePortForward(params[1].replace(/[\D]*/, '');
    }
  }
}
```

Comment (babyrjw) [inline lib/chromedriver.js]: Thanks for your comments. My honor to learn from your code.

File: lib/chromedriver.js [hunk 10]
```diff
@@ -226,12 +226,22 @@ class Chromedriver extends events.EventEmitter {
...
-      log.info("No old chromedrivers seemed to exist");
+      log.warn("No old chromedrivers seemed to exist");
+    }
+    log.debug(`Clean any old adb forwarded port socket connections`);
+    let connections = await this.adb.getForwardList();
+    for (let conn of connections) {
+      if (conn.indexOf('webview_devtools') !== -1) {
+        let params = conn.split(/\s+/);
+        if (params.length > 1) {
+          await this.adb.removePortForward(params[1].replace(/[\D]*/, '');
```
Comment (imurchie) [inline lib/chromedriver.js]: Missing closing parenthesis.

File: lib/chromedriver.js [hunk 11]
```diff
@@ -16,11 +16,11 @@ const DEFAULT_HOST = '127.0.0.1';
 const DEFAULT_PORT = 9515;
 class Chromedriver extends events.EventEmitter {
   constructor (args = {}) {
-    const {host, port, executable, cmdArgs, adbPort, verbose, logPath} = args;
+    const {host, port, executable, cmdArgs, adb, verbose, logPath} = args;
     super();
     this.proxyHost = host || DEFAULT_HOST;
     this.proxyPort = port || DEFAULT_PORT;
-    this.adbPort = adbPort;
+    this.adb = adb;
```
Comment (jlipps) [inline lib/chromedriver.js]: we should remember that this will be a breaking change, so we should publish this as a new major version of the package

Comment (jlipps) [inline lib/chromedriver.js]: or, if we don't want a breaking change, we could play nice, and allow both `adb` and `adbPort`. if the former is included we ignore the latter. if the latter is included but not the former, then we create a new `ADB` instance with the specified port, and assign it to `this.adb`.

Comment (imurchie) [inline lib/chromedriver.js]: We only use it in one place (all chrome handling is done through `appium-android-driver`) so it will be easy to handle the breaking change.

Comment (jlipps) [inline lib/chromedriver.js]: 👍

File: lib/chromedriver.js [hunk 12]
```diff
@@ -62,7 +62,7 @@ class Chromedriver extends events.EventEmitter {
     if (this.logPath) {
       args = args.concat([`--log-path=${this.logPath}`]);
     }
-
+
```
Comment (imurchie) [inline lib/chromedriver.js]: This added space is still a problem for linting. Did it pass locally?

Comment (babyrjw) [inline lib/chromedriver.js]: Yes, It did pass lint part of test locally. But It didn't pass unit-test. And I'm still finding the problem
image

Comment (babyrjw) [inline lib/chromedriver.js]: let me try to remove the spaces of this line

**Comment by imurchie** [comment]
You have made changes to the way we instantiate the ChromeDriver object. So the tests will need to be updated. The following is a diff that will fix them:
```
diff --git a/package.json b/package.json
index e7c3cc1..656e002 100644
--- a/package.json
+++ b/package.json
@@ -57,6 +57,7 @@
     "lint": "gulp eslint"
   },
   "devDependencies": {
+    "appium-adb": "^2.28.0",
     "appium-gulp-plugins": "^2.1.4",
     "babel-eslint": "^7.1.1",
     "chai": "^3.0.0",
diff --git a/test/chromedriver-specs.js b/test/chromedriver-specs.js
index 5adc56a..5287ed3 100644
--- a/test/chromedriver-specs.js
+++ b/test/chromedriver-specs.js
@@ -6,6 +6,8 @@ import chai from 'chai';
 import chaiAsPromised from 'chai-as-promised';
 import B from 'bluebird';
 import { exec } from 'teen_process';
+import ADB from 'appium-adb';
+
 
 let should = chai.should();
 chai.use(chaiAsPromised);
@@ -60,7 +62,9 @@ function buildReqRes (url, method, body) {
 describe('chromedriver binary setup', function () {
   this.timeout(20000);
   before(async () => {
-    let cd = new Chromedriver();
+    let cd = new Chromedriver({
+      adb: await ADB.createADB(),
+    });
     try {
       await cd.initChromedriverPath();
     } catch (err) {
@@ -81,7 +85,9 @@ describe('chromedriver with EventEmitter', function () {
   let cd = null;
   const caps = {browserName: 'chrome'};
   before(async () => {
-    let opts = {};
+    let opts = {
+      adb: await ADB.createADB(),
+    };
     cd = new Chromedriver(opts);
   });
   it('should start a session', async () => {
@@ -146,7 +152,10 @@ describe('chromedriver with EventEmitter', function () {
     await nextStatePromise.should.become(Chromedriver.STATE_STOPPED);
   });
   it('should throw an error when chromedriver doesnt exist', async () => {
-    let cd2 = new Chromedriver({executable: '/does/not/exist'});
+    let cd2 = new Chromedriver({
+      adb: await ADB.createADB(),
+      executable: '/does/not/exist',
+    });
     let nextErrP = nextError(cd2);
     await cd2.start({}).should.eventually.be.rejectedWith(/Trying to use/);
     let err = await nextErrP;
@@ -160,7 +169,9 @@ describe('chromedriver with async/await', function () {
   let cd = null;
   const caps = {browserName: 'chrome'};
   before(async () => {
-    let opts = {};
+    let opts = {
+      adb: await ADB.createADB(),
+    };
     cd = new Chromedriver(opts);
   });
   it('should start a session', async () => {
@@ -185,12 +196,17 @@ describe('chromedriver with async/await', function () {
     await assertNoRunningChromedrivers();
   });
   it('should throw an error during start if spawn does not work', async () => {
-    let badCd = new Chromedriver({port: 1});
+    let badCd = new Chromedriver({
+      adb: await ADB.createADB(),
+      port: 1,
+    });
     await badCd.start(caps).should.eventually.be.rejectedWith('ChromeDriver crashed during startup');
     await assertNoRunningChromedrivers();
   });
   it('should throw an error during start if session does not work', async () => {
-    let badCd = new Chromedriver();
+    let badCd = new Chromedriver({
+      adb: await ADB.createADB(),
+    });
     await badCd.start({chromeOptions: {badCap: 'foo'}})
                .should.eventually.be.rejectedWith('cannot parse capability');
     await assertNoRunningChromedrivers();
```

File: lib/chromedriver.js [hunk 13]
```diff
@@ -226,12 +225,22 @@ class Chromedriver extends events.EventEmitter {
     } else {
       cmd = `pkill -15 -f "${this.chromedriver}.*--port=${this.proxyPort}"`;
     }
-    log.info(`Killing any old chromedrivers, running: ${cmd}`);
+    log.debug(`Killing any old chromedrivers, running: ${cmd}`);
     try {
-      await (B.promisify(cp.exec))(cmd);
-      log.info("Successfully cleaned up old chromedrivers");
+      await B.promisify(cp.exec(cmd));
```
Comment (mykola-mokhnach) [inline lib/chromedriver.js]: why don't we use teen_process's exec here?

File: lib/chromedriver.js [hunk 14]
```diff
@@ -226,12 +225,22 @@ class Chromedriver extends events.EventEmitter {
...
-      log.info("No old chromedrivers seemed to exist");
+      log.warn("No old chromedrivers seemed to exist");
+    }
+    log.debug(`Clean any old adb forwarded port socket connections`);
+    let connections = await this.adb.getForwardList();
+    for (let conn of connections) {
+      if (conn.indexOf('webview_devtools') !== -1) {
+        let params = conn.split(/\s+/);
+        if (params.length > 1) {
+          await this.adb.removePortForward(params[1].replace(/[\D]*/, ''));
```
Comment (mykola-mokhnach) [inline lib/chromedriver.js]: `params[1].replace(/[\D]/g, '')` might be more stable

File: lib/chromedriver.js [hunk 15]
```diff
@@ -226,12 +225,22 @@ class Chromedriver extends events.EventEmitter {
...
+      await B.promisify(cp.exec(cmd));
+      log.debug("Successfully cleaned up old chromedrivers");
     } catch (err) {
-      log.info("No old chromedrivers seemed to exist");
+      log.warn("No old chromedrivers seemed to exist");
+    }
+    log.debug(`Clean any old adb forwarded port socket connections`);
+    let connections = await this.adb.getForwardList();
+    for (let conn of connections) {
+      if (conn.indexOf('webview_devtools') !== -1) {
```
Comment (mykola-mokhnach) [inline lib/chromedriver.js]: can we add a comment about why "webview_devtools" string is a magic constant here?

File: lib/chromedriver.js [hunk 16]
```diff
@@ -50,8 +50,8 @@ class Chromedriver extends events.EventEmitter {
     }
 
     let args = ["--url-base=wd/hub", `--port=${this.proxyPort}`];
-    if (this.adbPort) {
-      args = args.concat([`--adb-port=${this.adbPort}`]);
+    if (this.adb.adbPort) {
+      args = args.concat([`--adb-port=${this.adb.adbPort}`]);
```
Comment (mykola-mokhnach) [inline lib/chromedriver.js]: -> ```args.push(`--adb-port=${this.adb.adbPort}`)```

File: lib/chromedriver.js [hunk 17]
```diff
@@ -16,11 +16,11 @@ const DEFAULT_HOST = '127.0.0.1';
 const DEFAULT_PORT = 9515;
 class Chromedriver extends events.EventEmitter {
   constructor (args = {}) {
```
Comment (mykola-mokhnach) [inline lib/chromedriver.js]: args can be defined as `{host, port, executable, cmdArgs, adb, verbose, logPath}` right here

File: lib/chromedriver.js (hunk 13 above)
Comment (imurchie) [inline lib/chromedriver.js]: It doesn't handle pipes.

File: lib/chromedriver.js (hunk 16 above)
Comment (imurchie) [inline lib/chromedriver.js]: This should go in a different PR.

File: lib/chromedriver.js (hunk 17 above)
Comment (imurchie) [inline lib/chromedriver.js]: Again, outside the scope of this PR.

**Comment by imurchie** [comment]
Argh! My apologies for the confusion. I think to get CI working, rather than changing the tests to pass in adb, the use of adb within the code should gated by a check for its presence.

File: lib/chromedriver.js [hunk 18]
```diff
@@ -50,8 +50,8 @@ class Chromedriver extends events.EventEmitter {
     }
 
     let args = ["--url-base=wd/hub", `--port=${this.proxyPort}`];
-    if (this.adbPort) {
-      args = args.concat([`--adb-port=${this.adbPort}`]);
+    if (this.adb.adbPort) {
```
Comment (imurchie) [inline lib/chromedriver.js]: `if (this.adb && this.adb.adbPort) {`

File: lib/chromedriver.js [hunk 19]
```diff
@@ -226,12 +225,22 @@ class Chromedriver extends events.EventEmitter {
...
-      await (B.promisify(cp.exec))(cmd);
-      log.info("Successfully cleaned up old chromedrivers");
+      await B.promisify(cp.exec(cmd));
+      log.debug("Successfully cleaned up old chromedrivers");
     } catch (err) {
-      log.info("No old chromedrivers seemed to exist");
+      log.warn("No old chromedrivers seemed to exist");
+    }
+    log.debug(`Clean any old adb forwarded port socket connections`);
+    let connections = await this.adb.getForwardList();
```
Comment (imurchie) [inline lib/chromedriver.js]:
```js
if (this.adb) {
  let connections = ...
  // ...
}
```

File: lib/chromedriver.js (hunk 15 above)
Comment (babyrjw) [reply lib/chromedriver.js]: Yes, It's better for understanding

File: lib/chromedriver.js (hunk 14 above)
Comment (babyrjw) [reply lib/chromedriver.js]: It seems the params[1] can only be "tcp:port". So we strip the prefix. If ADB change the protocol in the future It should doesn't work. Because the protocol of removePortForward may also be change

File: test/chromedriver-specs.js [hunk 20]
```diff
@@ -6,6 +6,8 @@ import chai from 'chai';
 import chaiAsPromised from 'chai-as-promised';
 import B from 'bluebird';
 import { exec } from 'teen_process';
+import ADB from 'appium-adb';
```
Comment (imurchie) [inline test/chromedriver-specs.js]: These changes to the test are now unnecessary. And this is what is causing CI to fail, since Travis doesn't give access to the Android SDK on Mac.

Comment (imurchie) [reply test/chromedriver-specs.js]: this applies to the `package.json` addition, too. Sorry about that.

**Comment by imurchie** [comment]
I'm going to merge and do the minor fix in the tests. Thank you very much for the work! It is great to have this be better functioning!

**Comment by imurchie** [comment]
Published in `appium-chromedriver@2.12.3`

**Comment by imurchie** [comment]
Sorry. Forgot this was a breaking change.

Published in `appium-chromedriver@3.0.0`.

**Comment by babyrjw** [comment]
Thanks for your patience

**Comment by jlipps** [comment]
> Published in appium-chromedriver@3.0.0

Thanks @imurchie. Just a question: did you unpublish the earlier one?

**Comment by imurchie** [comment]
@jlipps Yes. And re-published a new version of the old version (i.e., without the change). So the latest in the 2.x line is correct.

**Comment by jlipps** [comment]
Excellent!
